jni/src/main/java/com/horizen/storage/Storage.java

```diff
@@ -0,0 +1,137 @@
+package com.horizen.storage;
+
+import com.horizen.common.ColumnFamily;
+import com.horizen.common.DBIterator;
+import com.horizen.common.interfaces.ColumnFamilyManager;
+import com.horizen.common.interfaces.DefaultReader;
+import com.horizen.librust.Library;
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+public class Storage implements DefaultReader, ColumnFamilyManager, AutoCloseable {
+    // Loading the Rust library which contains all the underlying logic
+    static {
+        Library.load();
+    }
+
+    private long storagePointer;
+    final private ColumnFamily defaultCf;
+
+    public void checkPointer() throws IllegalStateException {
+        if (storagePointer == 0)
+            throw new IllegalStateException("Storage instance was freed");
+    }
+
+    // Constructor is intended to be called from inside the Rust environment for setting a raw pointer to a Rust-instance of Storage
+    private Storage(long storagePointer, long defaultColumnFamilyPointer) {
+        this.storagePointer = storagePointer;
+        this.defaultCf = new ColumnFamily(defaultColumnFamilyPointer);
```

Comment:
Yes, this option is also possible and was considered. But creation of the `ColumnFamily` object with the `create_java_object` method entails an additional call to JVM from the native Rust context. So to avoid this the creation of the `ColumnFamily` object is moved to the Java side: when calling the `StorageVersioned`/`TransactionVersioned` Java-constructor 
 a pointer of the `ColumnFamily`'s Rust object is passed together with a pointer of a storage/transaction Rust object thus a ColumnFamily Java object can be created in this constructor on Java side. So this approach is used just for some optimization, because instead of two calls to JVM (creation of the `ColumnFamily` Java-object and subsequent creation of the `StorageVersioned`/`TransactionVersioned` Java object both from the Rust side) we make a one call to JVM (a `ColumnFamily` Java-object is created on the Java side while calling the `StorageVersioned`/`TransactionVersioned` constructor from the Rust side). But yes, then we need to make `ColumnFamily`'s Java constructor public (it's a pity that Java doesn't support the `friend class` semantics like C++ does). But if it is important to keep the Java `ColumnFamily`'s constructor private, then I can change the implementation to 'ColumnFamily' creation with `create_java_object`.